contracts/Lib/DmpAlgorithm.sol

```diff
@@ -0,0 +1,85 @@
+/*
+Copyright: Ambrosus Inc.
+Email: tech@ambrosus.com
+
+This Source Code Form is subject to the terms of the Mozilla Public License, v. 2.0. If a copy of the MPL was not distributed with this file, You can obtain one at https://mozilla.org/MPL/2.0/.
+
+This Source Code Form is “Incompatible With Secondary Licenses”, as defined by the Mozilla Public License, v. 2.0.
+*/
+
+pragma solidity ^0.4.23;
+
+import "openzeppelin-solidity/contracts/math/SafeMath.sol";
+
+import "./SafeMathExtensions.sol";
+
+
+library DmpAlgorithm {
+    using SafeMath for uint;
+    using SafeMath for uint32;
+    using SafeMath for uint64;
+    using SafeMathExtensions for uint;
+
+    uint constant public DMP_PRECISION = 100;
+
+
+    function qualifyShelterer(bytes32 dmpBaseHash, uint dmpLength, uint currentRound) internal pure returns (uint32) {
+        return uint32(uint256(keccak256(abi.encodePacked(dmpBaseHash, currentRound))).mod(dmpLength));
+    }
+
+    function selectingAtlasTier(bytes32 dmpBaseHash, uint[] atlasCount, uint[] ATLAS_NUMERATOR) internal pure returns (uint32) {
+        require(atlasCount.length == ATLAS_NUMERATOR.length);
+        
+        uint[] memory wcd = calculateWcd(atlasCount, ATLAS_NUMERATOR);
+
+        uint32 tier = selectRandomlyFrom(dmpBaseHash, wcd);
+
+        return tier;
+    }
+
+    function calculateDenominator(uint[] atlasCount, uint[] ATLAS_NUMERATOR) private pure returns (uint) {
+        uint denominator = 0;
+        for (uint tier = 0; tier < ATLAS_NUMERATOR.length; tier++) {
+            uint elem = atlasCount[tier].mul(ATLAS_NUMERATOR[tier]);
+            denominator = denominator.add(elem);
+        }
+        return denominator;
+    }
+
+    function calculateWcd(uint[] atlasCount, uint[] ATLAS_NUMERATOR) private pure returns (uint[]) {
+        uint[] memory wcd = new uint[](ATLAS_NUMERATOR.length);
+        uint currentWCD = 0;
+        uint denominator = calculateDenominator(atlasCount, ATLAS_NUMERATOR);
+
+        for (uint tier = 0; tier < ATLAS_NUMERATOR.length; tier++) {
+            uint currentNum = atlasCount[tier].mul(ATLAS_NUMERATOR[tier]);
+
+            if (currentNum == denominator) {
```

Comment:
@Mike-S-Mike I think we can get rid of this `if` and leave only the part in `else`. Am I missing some special edge case or do you agrees?

Reply:
This case is needed if we only have one type of nodes, or if there are no nodes of a certain type. Without this we can get probability 99% (e.g. due to rounding errors) instead of 100. So, in that case if we roll 100 on the DMP, the algorithm won't be able to select anything and will fail.

Reply:
But then there's the correction for that case in lines 67-72, isn't it?

Reply:
Yes, you are right, 67-72 lines have been added later than 'If' statement and they resolve such situation. 'If' statement was removed.